### PR TITLE
Add rank-local determinism tracing

### PR DESCRIPTION
**Pull request by junjiewang253-ctrl**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Adds an opt-in, rank-local determinism tracer and offline comparator for locating the first semantic divergence between distributed training runs.

The implementation follows the progressive workflow discussed in the determinism roadmap:

1. record low-cost lifecycle and tensor metadata on selected ranks/iterations;
2. narrow the first mismatching boundary with summary or deterministic sampled digests;
3. use full-tensor SHA-256 only around the localized boundary.

The tracer writes one append-only JSONL file per global rank and never introduces a distributed collective. It is disabled by default.

## Issue tracking

Linked issue: Relates to #5785

Design discussion: https://github.com/NVIDIA/Megatron-LM/issues/5785#issuecomment-5323286193

## Implementation
- Adds training flags for output directory, rank/iteration selectors, evidence mode, sampling, flushing, and explicit append behavior.
- Keeps the reusable tracing recorder and schema under megatron/core/determinism, with megatron/training limited to lifecycle and metric integration.
- Records iteration lifecycle, scalar losses, gradient norm, and zero-gradient count from the training loop.
- Supports metadata, summary, sampled, and full evidence modes.
- Uses create-only output by default; append mode validates the existing stream and resumes sequence/occurrence counters.
- Adds an offline comparator that reports the earliest divergence independently per rank in execution order, without implying a global cross-rank ordering.
- Rejects malformed rank trace filenames and invalid trace records fail-closed.
- Adds CUDA coverage for non-default streams, observer validity, and CUDA graph restrictions.
- Keeps backend/model-specific boundary sites and the deferred device-digest backend out of this PR for follow-up work.

## Semantics and safety

- No distributed collectives are called by the tracer.
- Metadata mode does not read tensor values.
- Value-bearing modes snapshot tensors at the call site and materialize them at flush time.
- Sampled and summary matches are diagnostic only.
- A `full_tensor_certificate` means every captured tensor record hashed every logical tensor byte; it does not claim the uncaptured program state was equal.
- Existing trace files are never overwritten unless append continuation is explicitly requested.
- Pending records are flushed at iteration boundaries, graceful shutdown, and training exception handling. SIGKILL remains inherently unrecoverable.

## Validation
 - [x] I have added relevant unit tests
 - [x] I have added relevant functional tests
- CUDA tests cover non-default streams, observer validity, and CUDA graph behavior.
- The 14 CUDA tests are currently skipped locally because GPU access is unavailable.
 - [x] I have added proper typing to my code
 - [x] I have added relevant documentation
 - [] I have run tools/autoformat.sh on my PR

Local checks:

- 96 CPU determinism-trace contract tests passed.
- 14 CUDA tests were collected and skipped because no CUDA device is available.
- Black, isort, Ruff, and pylint passed for the scoped changes.
- git diff --check passed for the PR diff.

Target-stack GPU integration and observer-performance qualification remain pending.

## Remaining validation
- Run the CUDA test suite on the target GPU stack.
- Measure observer overhead for the supported evidence modes.
- Run the repository-standard CI/formatting environment when available.
- Keep the deferred device-digest backend and backend-specific FSDP/optimizer/collective/TE boundaries as follow-up work, per maintainer discussion.

### CODE REVIEW

**Comment by Connor-XY** [comment]
Thanks for implementing this and for keeping the initial scope generic. We reviewed the direction in the context of #5785 and would like to treat #6610 as a candidate common foundation, rather than build a separate public tracing system.

The rank-local schema, selectors, append behavior, evidence-strength distinctions, documentation, and fail-closed comparator are all useful. We can contribute complementary work from the ongoing hierarchical-tracing effort: a preallocated device-digest backend with deferred host materialization, targeted FSDP/optimizer boundaries, and GPU-scale validation.

Before detailed line-by-line review, we would like to align on four points:

1. **Layering:** please keep the reusable recorder/digest API in a Megatron-Core-safe module, with only lifecycle and metric wiring under `megatron/training`. Future Core boundary sites should not need to import a training-layer module.
2. **First-divergence semantics:** the comparator currently validates `sequence`, but selects content mismatches using semantic-key sorting. Within an iteration this can report the lexicographically first boundary rather than the first one executed. “First” should mean earliest per-rank execution sequence. Across ranks, the report should expose the earliest candidate for each rank rather than imply a synchronized global order.
3. **Observer strength:** `summary`, `sampled`, and `full` add device work, allocation, and host synchronization even though they add no collectives. That is acceptable for progressive narrowing, but the behavior and intended dose should remain explicit. We propose adding a deferred device-digest mode in a follow-up for low-perturbation scale localization, while retaining full SHA-256 for narrow confirmation.
4. **Validation:** CPU contract tests are a good start. Before treating value-bearing captures as scale-safe, we should add GPU tests for non-default streams and an observer-validity/performance check. We can help run those tests on the target stack.

We suggest keeping backend/model-specific call sites out of this PR. Once the common API and comparison semantics are settled, we can submit FSDP, collective, Transformer Engine, and optimizer boundaries as smaller follow-ups.

There is no need to close or restart this work. If this division sounds reasonable, we can share the overlapping implementation and GPU findings, agree on the recorder/backend interface, and collaborate on the changes directly in or following #6610.

**Comment by junjiewang253-ctrl** [comment]
Thanks, this makes sense to me.

I agree with keeping the reusable recorder/digest pieces Core-safe, and leaving megatron/training responsible only for the training lifecycle integration.

I’ll also update the comparator so that “first divergence” is based on the actual per-rank execution sequence, rather than implying a global ordering across ranks.

For the evidence modes, I’ll keep the current metadata/summary/sampled/full distinction, but I’m happy to align the backend interface with the deferred device-digest work you already have instead of introducing a separate abstraction.

I’ll keep FSDP/optimizer/collective/TE-specific boundaries out of this PR and treat those as follow-ups.

I’ll first clean up the current review issues and the layering/comparator semantics. If you can share the interface shape or constraints from the overlapping implementation, that would be very helpful before I do the larger refactor.

And thanks for offering help with the GPU validation — non-default-stream coverage and observer-overhead checks sound like the right next validation steps.

**Comment by junjiewang253-ctrl** [comment]
I pushed the first revision based on the points above. The reusable recorder/schema now live under megatron/core/determinism, while training only contains lifecycle integration. The comparator now reports the earliest divergence independently per rank in execution sequence order, and the trace loader now fails closed on malformed rank filenames.

I also added CUDA tests for non-default streams/observer validity and made sampled capture avoid full flattening of non-contiguous tensors. Locally, 96 CPU contract tests pass; the 14 CUDA tests are currently skipped because I don't have GPU access.

The branch is now synced with current main. I’ve left the device-digest backend and backend-specific boundary sites out of this revision as discussed.

File: tools/determinism/trace_comparison.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    root = Path(path)
+    records: list[dict[str, Any]] = []
+    ranks_seen: set[int] = set()
+    for trace_file in _trace_files(root):
+        match = _RANK_FILE_RE.search(trace_file.name)
+        expected_rank = int(match.group(1)) if match else None
+        if expected_rank is not None and expected_rank in ranks_seen:
+            raise TraceValidationError(f"duplicate trace file for rank {expected_rank}")
+        if expected_rank is not None:
+            ranks_seen.add(expected_rank)
```
[the remark this answers is not in the document]
Comment (junjiewang253-ctrl) [reply tools/determinism/trace_comparison.py]: Fixed in the latest revision — directory loads now require rank_<digits>.jsonl and fail closed otherwise.
